lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -139,95 +150,275 @@
           }
         }
 '''
+import json
+
+from abc import ABCMeta, abstractmethod
+
+from ansible.module_utils._text import to_native, to_text
+from ansible.module_utils.basic import AnsibleModule, get_all_subclasses
+from ansible.module_utils.six import with_metaclass
+
+
+class TestFailed(Exception):
+    pass
+
+
+class PkgMgr(with_metaclass(ABCMeta, object)):
+
+    def __init__(self):
+
+        if not self.test():
+            raise TestFailed("Test failed to pass")
+
+    @abstractmethod
+    def test(self):
+        ''' This method is supposed to return True/False if the package manager is installed/usable by this module '''
+        pass
+
+    @abstractmethod
+    def list_installed(self):
+        ''' This method should return a list of installed packages, each list item will be passed to get_package_details '''
+        pass
+
+    @abstractmethod
+    def get_package_details(self, package):
+        ''' This takes a 'package' item and returns a dictionary with the package information, name and version are minimal requirements '''
+        pass
+
+    def get_packages(self):
+        ''' Take all of the above and return a dictionary of lists of dictionaries (package = list of installed versions) '''
+
+        installed_packages = {}
```

Comment:
Please use:
```suggestion
        installed_packages = defaultdict(list)
```

This way you'll simplify logic below.